File: cplusplus/stack/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        const map<char, char>pairs{ {')', '('},  {'}', '{'}, {']', '['} };
        stack<char> tracker{};

        if (s.length() == 1) return false;

        for (auto subString : s) {
            if (subString == '(' || subString == '{' || subString == '[') {
                tracker.push(subString);
            }
            else if (!tracker.empty()) {
                if (pairs.at(subString) != tracker.top()) {
                    return false;
                }
                tracker.pop();
            }
            else {
                return false;
            }
        }

        if (!tracker.empty()) {
            return false;
        }
        else {
            return true;
        }
    }
};

//https://leetcode.com/problems/valid-parentheses

File: python/stack/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:

        if len(s) % 2 != 0:
            return False

        bracketMap = {']': '[', ')': '(', '}': '{'}
        bracketTracker = []

        for i in s:
            if i in { '(', '[', '{' }:
                bracketTracker.append(i)
            elif len(bracketTracker) != 0 and bracketTracker[-1] == bracketMap[i]:
                bracketTracker.pop()
            else:
                return False

        return False if len(bracketTracker) != 0 else True


# https://leetcode.com/problems/valid-parentheses/
        
        